fms_extras/utils/generation.py

```diff
@@ -0,0 +1,156 @@
+        cudagraphs: bool
+            if True, model is using cudagraphs and input will be padded, otherwise no cudagraphs is used and input is
+            not padded
+
+    Returns:
+    Tuple[torch.Tensor, int, float, float]
+        the resulting output tokens, the number of new tokens generated, the time to first token in seconds and the
+        cumulative time of all decode steps in seconds
+    """
+    start_time = time.time()
+    if decode_model is None:
+        decode_model = model
+
+    bsize = len(input_ids_list)
+
+    if cudagraphs and bsize != 1:
+        raise NotImplementedError(
+            "cudagraphs is not yet supported for batch sizes greater than 1"
+        )
+
+    # Build padded batched input tensor
+    max_len = max([seq.size(0) for seq in input_ids_list])
+    model_input_lengths = [seq.size(0) for seq in input_ids_list]
+    input_ids = torch.stack(
+        [
+            F.pad(input_ids_list[i], (max_len - model_input_lengths[i], 0))
+            for i in range(bsize)
+        ]
+    )
+
+    result = input_ids
+    next_input = input_ids
+    kwargs: MutableMapping[str, Any] = dict()
+    kwargs["cache_data"] = None
+    kwargs["use_cache"] = True
+    kwargs["mask"] = __create_prefill_mask(model_input_lengths, device=input_ids.device)
+    sequence_ids: Optional[List[int]] = None
+    block_mapping_max = ((max_len + max_new_tokens) // 16) + 1
+    for i in range(max_new_tokens):
+        if i == 1:
+            start_time = time.time()
+            kwargs["mask"] = None
+            sequence_ids = kwargs["cache_data"].sequence_ids
+            model_input_lengths = [1 for _ in range(bsize)]
+
+        input_ids = next_input[:, -max_seq_len:]
+
+        kwargs["cache_data"] = kv_cache_manager.allocate_tokens(
+            model_input_lengths, sequence_ids
+        )
+
+        if i == 0:
+            logits, _ = model(input_ids, **kwargs)
+        else:
+            # cudagraph requires static shapes
+            if cudagraphs:
+                kwargs["cache_data"].block_mapping = __right_pad_zeros(
+                    kwargs["cache_data"].block_mapping, block_mapping_max
+                )
+
+            logits, _ = decode_model(input_ids, **kwargs)
+        logits = logits[:, -1, :]
+
+        if do_sample:
+            # get logits from last value in sequence nad scale
+            logits = logits / temperature
+            if top_k:
+                v, _ = torch.topk(logits, top_k)
+                logits[logits < v[:, [-1]]] = -float("inf")
+
+            probs = F.softmax(logits, dim=-1)
+            next_val = torch.multinomial(probs, num_samples=1)
+        else:
+            next_val = torch.argmax(logits, dim=-1).unsqueeze(0).t()
+
+        result = torch.cat((result, next_val), dim=-1)
+
+        next_input = next_val
+
+        if i == 0:
+            ttft = time.time() - start_time
```

Comment:
also it might be easier to follow if we move the `start_time = time.time()` to here, right after setting ttft, instead of under `if i == 1`.